aten/src/ATen/native/README.md

```diff
@@ -268,26 +268,63 @@ dispatch:
     registered to `Math` key should be plain mathematical composition of other
     `at::` functions and support training and inference for any backend.
 
-If you add `dispatch` section to any API that didn't have it before, you **have to** move
-the old implementation to `Math` field so that it's still available for other backends to use.
+`DefaultBackend` and `Math` keys act as defaults: for example, you can specify a custom
+kernel for a particular backend using a backend-specific dispatch key, and use
+`DefaultBackend` or `Math` to specify a generic kernel for the others.
+
+Note that like those registered to `Math`, kernels registered to `DefaultBackend` are
+very often implemented as mathematical expressions built up from calls to other `at::`
+functions.  This is because in both cases, the kernel needs to delegate backend-specific
+computation to the functions it calls.  The difference between `DefaultBackend` and `Math`
+is that a `Math` kernel also implicitly defines a derivative formula: to do this, it must
+call only functions that themselves support autograd.
+
+For example, suppose `my_op` can be implemented in the following way:
+```
+at::Tensor my_op(const Tensor& self, const Tensor& other) {
+  return self + 2 * other;
+}
+```
+
+If we already know inference kernels and derivative formulas for operators `+` and `*` in our system,
+you can just register `my_op` to `Math` and both inference & autograd will just work.
+Although it seems we only write down the inference formula here, PyTorch autograd system would correctly
+set up the backward for `my_op` using the chain formula and derivatives of `+` & `-` operators.
+In other words `d_out/d_self = 1; d_out/d_other = 2` can be derived automatically from
+the `my_op` inference kernel. Of course if we don't have derivative formula defined for either `+` or `-`,
+backward of `my_op` can no longer be derived automatically.
+
+Whether to use `Math` or `DefaultBackend` for your kernel can be decided by the following steps:
+1. If you can, always start with a `Math` kernel that's composable from existing operators.
+2. If you don't want to use the derived gradient formula from `Math` kernel for autograd, either to
+   get better performance or better numerical stability, you should put the kernel in `DefaultBackend`
+   so that it's only used in inference.
+   Later for autograd, depending on whether your autograd kernel works for all backends or not,
+   you can put them in alias `Autograd` or specific keys like `AutogradCPU`.
+3. If you prefer to write backend-specific kernels, use reserved dispatch keys for your backend instead,
+   e.g. `CPU/AutogradCPU`.
+
+If you add backend specific kernel to any operator that only had `Math` kernel before(no dispatch section
+or only `Math` entry in dispatch section), you **have to** move the old implementation to a `<op>_math`
+kernel and put it in `Math` field so that it's still available for other backends to use.
```

Comment:
I don't think we actually require a name change though, right? Sorry about my confusing  earlier comment btw - I was thinking about telling them how to reconstruct the name of the old implementation, but of course they can just look it up. How about this for the whole paragraph:

---

**Important**: because a `Math` kernel is implicitly registered for ops with no `dispatch:` section, when you add a backend-specific kernel (and hence a `dispatch:` section) to one of these, you **must** also
add a `Math:` entry that names the old kernel implementation (it's named after the op, with `_<overload name>` added if applicable), so that it's still available for other backends to use. 

---

(I was going to put something about adding a `_math` suffix for clarity, but I just checked native_functions and we don't actually do that! Most `Math:` kernels are just the op name, with a couple that put `math_` in the _front_. So maybe we shouldn't suggest anything in the docs, and just let people take their cues from what they find in native_functions. (If you think it's important for clarity we could say something here, but then we should codemod to make native_functions match, I'd say)

Reply:
SGTM! Thanks for the suggestion!